<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html xmlns:th="http://www.thymeleaf.org" lang="en-us">
<head>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-139406062-1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'UA-139406062-1');
    </script>

    <title>ORFanGenes</title>
    <link rel="shortcut icon" type="image/x-icon" href="assets/favicon.ico" />
    <!--Import Google Icon Font-->
    <link href="http://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
    <!--Import materialize.css-->
    <link type="text/css" rel="stylesheet" href="assets/css/materialize.min.css" media="screen,projection" />
    <!--Let browser know website is optimized for mobile-->
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="csrf-token" content="{{ csrf_token() }}">
    <!--Import jQuery before materialize.js-->
    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
    <script type="text/javascript" src="assets/js/materialize.min.js"></script>
    <script type="text/javascript" src="assets/js/orfanid-clamp.js"></script>
</head>
<body>
<div th:replace="fragments/header :: nav">&nbsp;</div>
<main>
    <form action="/clamp" th:action="@{/clamp}" th:object="${chromosome}" method="post">
    <div class="container">
        <div class="row">
            <h4>Comparative Genomic Studies of ORFan genes in Mammalian Genomes
            </h4>
        </div>

        <div class="row">
            <div class="col s6">
                <h5>Purpose</h5>
                <p>Investigate the function/s, if any, of the ORFan gene sequences identified in previous studies from the literature, with a special focus of Clamp at el study. However, the initial goal is to find out all the orphan genes of the human genome.</p>

                <h5>Methods</h5>
                <p>Human Genome GRCh38.p7 was used for the analysis. </p>

                <p>A bioinformatics pipeline was developed using R language with BiomaRt package to filter genes. BiomaRt is the official filtering tool to query ensemble databases. There were six main steps in the filtering process:
                <ol>
                    <li>Removal of Retrotransposons/Pseudogenes</li>
                    <li>Removal of orthologous genes with Dog</li>
                    <li>Removal of orthologous genes with Mouse</li>
                    <li>Removal of paralogous genes within Human</li>
                    <li>Removal of known pfam genes</li>
                    <li>Removal of genes that absent the protein sequence (i.e. non-coding genes)</li>
                </ol>
                </p>
            </div>
            <div class="col s6">
                <img src="assets/images/Clamp.png" alt="Clamp method" height="50%" width="50%">
            </div>
        </div>
        <div class="row">
            <div class="col offset-s9 col s3">
                <button class="btn waves-effect waves-light" type="submit" name="action" id="submit">View Clamp Results
                    <i class="material-icons right">send</i>
                </button>
            </div>
        </div>
    </div>
    </form>
</main>
<div th:replace="fragments/footer :: footer">&nbsp;</div>
</body>
</html>